enzyme/Enzyme/GradientUtils.cpp

```diff
@@ -442,22 +442,36 @@ Value *GradientUtils::unwrapM(Value *const val, IRBuilder<> &BuilderM,
     assert(val->getType() == toreturn->getType());
     return toreturn;
   } else if (auto op = dyn_cast<InsertValueInst>(val)) {
-    auto op0 = getOp(op->getAggregateOperand());
-    if (op0 == nullptr)
-      goto endCheck;
-    auto op1 = getOp(op->getInsertedValueOperand());
-    if (op1 == nullptr)
-      goto endCheck;
-    auto toreturn = BuilderM.CreateInsertValue(op0, op1, op->getIndices(),
-                                               op->getName() + "_unwrap");
-    if (permitCache)
-      unwrap_cache[BuilderM.GetInsertBlock()][idx.first][idx.second] = toreturn;
-    if (auto newi = dyn_cast<Instruction>(toreturn)) {
-      newi->copyIRFlags(op);
-      unwrappedLoads[newi] = val;
-      if (newi->getParent()->getParent() != op->getParent()->getParent())
-        newi->setDebugLoc(nullptr);
+    SmallVector<std::pair<Value *, unsigned>, 0> insertElements;
+    Value *agg = op->getAggregateOperand();
+    while (auto op1 = dyn_cast<InsertValueInst>(agg)) {
+      if (knownRecomputeHeuristic.count(op1) &&
+          knownRecomputeHeuristic[op1] == true)
+        break;
+      insertElements.push_back({op1->getInsertedValueOperand(),
```

Comment:
You should check that there's only 1 such index?

Reply:
or maybe just assert?

Reply:
No because there are plenty of multi-index insertions, and we would then have compilation fail when unwrapping them (ignorant of vector mode even)

Reply:
added check

Reply:
I don't see the check?

Reply:
forgot to commit and push

Reply:
Still dont see this `if (op1->getNumIndices() != 1) break` check

Reply:
The count doesn't need to check against 1, if that's what you misunderstood this as.